File: python/leaderboard_storage.py
# 리더보드 파일 저장소. 서비스 규칙과 분리된 동기식 저장 계약이다.
# Copyright 2026 HJOW, Apache License 2.0

import json
import threading
from pathlib import Path
from typing import Any


class FileLeaderboardStorage:
	"""생성만으로 파일에 접근하지 않는다. 초기화는 리더보드 기능을 켠 서비스가 요청한다.

	한 사람의 기록을 [홈디렉토리]/.puyowserver/leaderboard/[닉네임].json 한 파일에 모아 둔다.
	파이썬 서버는 ThreadingHTTPServer 라 요청마다 스레드가 다르므로, 읽고 고쳐 쓰는 묶음이
	서로 끼어들지 않도록 threading.Lock 을 둔다. (node.js 쪽은 단일 스레드 + 동기 입출력이라
	node/leaderboard_storage.js 에는 잠금이 없다.)

	닉네임 검증과 기록 형식 검사, 순위 병합은 서비스(leaderboard.py)가 담당한다.
	"""

	def __init__(self, root: Path | None = None) -> None:
		"""기존 홈 경로를 기본값으로 사용하며 테스트에서는 독립 경로를 주입한다."""
		storage_root = Path(root) if root is not None else Path.home() / ".puyowserver"
		self.leaderboard_root = storage_root / "leaderboard"
		self._lock = threading.Lock()

	def initialize(self) -> None:
		"""저장 디렉터리를 만든다. 실패는 호출자에게 전달한다."""
		self.leaderboard_root.mkdir(parents=True, exist_ok=True)

	def lock(self) -> threading.Lock:
		"""읽고 고쳐 쓰는 묶음을 감쌀 잠금이다. 서비스가 with 문으로 사용한다."""
		return self._lock

	def load_player(self, nickname: str) -> dict[str, Any] | None:
		"""한 사람의 기록 파일을 읽는다. 없거나 읽지 못하면 None이다."""
		return self._read_json_file(self.get_player_file_path(nickname))

	def save_player(self, player: dict[str, Any]) -> None:
		"""한 사람의 기록 파일을 통째로 저장한다. 실패는 호출자에게 전달한다."""
		self._write_json_file(self.get_player_file_path(str(player["nickname"])), player)

	def list_players(self) -> list[dict[str, Any]]:
		"""저장된 사람들의 기록 파일을 모두 읽는다. 손상되거나 읽을 수 없는 파일은 건너뛴다."""
		try:
			entries = sorted(self.leaderboard_root.glob("*.json"))
		except OSError:
			return []
		players: list[dict[str, Any]] = []
		for entry in entries:
			player = self._read_json_file(entry)
			if player and isinstance(player.get("nickname"), str):
				players.append(player)

		return players

	def get_player_file_path(self, nickname: str) -> Path:
		"""닉네임에 해당하는 파일 경로다. 닉네임 검증은 서비스가 이미 끝냈다고 본다."""
		return self.leaderboard_root / f"{nickname}.json"

	def _read_json_file(self, file_path: Path) -> dict[str, Any] | None:
		"""기록 파일을 읽는다. 파일이 없거나 형식이 깨졌으면 None을 돌려준다."""
		try:
			with file_path.open("r", encoding="utf-8") as stream:
				value = json.load(stream)
			return value if isinstance(value, dict) else None
		except (OSError, json.JSONDecodeError):
			return None

	def _write_json_file(self, file_path: Path, value: dict[str, Any]) -> None:
		"""부모 디렉터리를 만든 뒤 UTF-8 JSON으로 저장한다."""
		file_path.parent.mkdir(parents=True, exist_ok=True)
		with file_path.open("w", encoding="utf-8") as stream:
			json.dump(value, stream, ensure_ascii=False, indent=2)

File: node/leaderboard_storage.js
/** 리더보드 파일 저장소. 서비스 규칙과 분리된 동기식 저장 계약이다. @license Apache-2.0 */
const fs = require('fs');
const os = require('os');
const path = require('path');

/**
 * 기본 JSON 저장소. 생성만으로는 파일을 만들거나 지우지 않는다.
 * 다른 저장소도 initialize/loadPlayer/savePlayer/listPlayers 를 구현하면 그대로 갈아 끼울 수 있다.
 * 닉네임 검증과 기록 형식 검사, 순위 병합은 서비스(leaderboard.js)가 담당한다.
 *
 * 한 사람의 기록을 [홈디렉토리]/.puyowserver/leaderboard/[닉네임].json 한 파일에 모아 둔다.
 * 파일 입출력은 모두 동기 함수를 쓴다. node.js 는 단일 스레드이므로 읽기·고치기·쓰기 한 묶음이
 * 다른 요청에 끼어들지 않고 끝나며, 이것이 곧 이 저장소의 잠금 장치다. (파이썬 쪽은 스레드 서버라
 * python/leaderboard_storage.py 에서 threading.Lock 을 따로 쓴다.)
 */
class FileLeaderboardStorage {
    /** @param {string} root 기존 저장 경로 또는 테스트용 독립 경로 */
    constructor(root = path.join(os.homedir(), '.puyowserver')) {
        this.leaderboardRoot = path.join(root, 'leaderboard');
    }

    /** 리더보드 기능을 켰을 때만 서비스가 호출한다. 실패는 호출자에게 전달한다. */
    initialize() {
        fs.mkdirSync(this.leaderboardRoot, { recursive: true });
    }

    /**
     * 한 사람의 기록 파일을 읽는다. 없거나 읽지 못하면 null 이다.
     * @param {string} nickname 검증을 마친 닉네임
     * @returns {object|null} 저장된 객체
     */
    loadPlayer(nickname) {
        return this.readJsonFile(this.getPlayerFilePath(nickname));
    }

    /**
     * 한 사람의 기록 파일을 통째로 저장한다. 실패는 호출자에게 전달한다.
     * @param {object} player 저장할 객체
     * @returns {void}
     */
    savePlayer(player) {
        this.writeJsonFile(this.getPlayerFilePath(player.nickname), player);
    }

    /**
     * 저장된 사람들의 기록 파일을 모두 읽는다. 손상되거나 읽을 수 없는 파일은 건너뛴다.
     * @returns {object[]} 저장된 객체 목록
     */
    listPlayers() {
        let names = [];
        try {
            names = fs.readdirSync(this.leaderboardRoot);
        } catch {
            return [];
        }
        const players = [];
        names.forEach((name) => {
            if (!name.endsWith('.json')) return;
            const player = this.readJsonFile(path.join(this.leaderboardRoot, name));
            if (player && typeof player.nickname === 'string') players.push(player);
        });
        return players;
    }

    /**
     * 닉네임에 해당하는 파일 경로다. 닉네임 검증은 서비스가 이미 끝냈다고 본다.
     * @param {string} nickname 닉네임
     * @returns {string} 파일 경로
     */
    getPlayerFilePath(nickname) {
        return path.join(this.leaderboardRoot, `${nickname}.json`);
    }

    /**
     * JSON 파일을 읽어 객체로 돌려준다. 읽지 못하면 null 이다.
     * @param {string} filePath 파일 경로
     * @returns {object|null} 파싱한 객체
     */
    readJsonFile(filePath) {
        try {
            const value = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
            return value && typeof value === 'object' && !Array.isArray(value) ? value : null;
        } catch {
            return null;
        }
    }

    /**
     * 객체를 JSON 파일로 저장한다.
     * @param {string} filePath 파일 경로
     * @param {object} value 저장할 객체
     * @returns {void}
     */
    writeJsonFile(filePath, value) {
        fs.mkdirSync(path.dirname(filePath), { recursive: true });
        fs.writeFileSync(filePath, JSON.stringify(value, null, 2), 'utf-8');
    }
}

module.exports = { FileLeaderboardStorage };
